pyproject.toml

```diff
@@ -15,13 +15,18 @@ authors = [
 keywords = []
 dependencies = [
     "numpy",
-    "xarray",
-    "rioxarray",
-    "rasterio",
     "dask[diagnostics,dataframe]",
     "dask-ml>=2024.3.20",
     "scikit-learn",
+]
+
+[project.optional-dependencies]
+datasets = [
     "sknnr",
+    "xarray",
+    "rioxarray", 
+    "rasterio", 
+    "pooch",
```

Comment:
> Do you think that it's wise at this junction to disassociate sknnr-spatial from the sknnr "ecosystem"? The way you are designing it seems like it could serve any sklearn estimator and maybe sknnr-spatial just has the extra bits associated with neighbors and distances?? Obviously, that gets to be a lot of different packages to manage ...

Great question that I've been wondering about, too. Correct me if I'm wrong, but even when it comes to neighbors and distances, I'm not sure there's anything we'll need that won't also apply to a standard `KNeighborsRegressor`, right? Some of the differences between that and `sknnr` estimators like `return_dataframe_index` could be handled as kwargs. There might be exceptions I haven't thought about, though...

I do think there's a need for a general purpose sklearn-raster interface, so it probably would make sense to disassociate this from `sknnr` if it ends up being more broadly usable. If there is functionality that's specific to `sknnr`, maybe it's something we could tuck away in a module? 

If you're okay punting on this for now, I'd say we keep it in mind and revisit when we're a little further along and have a better sense of how compatible things are.

> For rioxarray, xarray, and rasterio, I think what you're saying is that eventually we will use these packages for prediction, but because that code hasn't yet been written, you want to keep the dependencies tied to what is currently present, correct?

I think that'll probably be the case for `xarray`, but `rioxarray` and `rasterio` would probably only be needed if we add functionality for reading/writing rasters. Not sure if that's something you had in mind as part of `sknnr-spatial`, or if that's something we leave up to users? In any case, not something we need to make a decision on now!

Reply:
> Correct me if I'm wrong, but even when it comes to neighbors and distances, I'm not sure there's anything we'll need that won't also apply to a standard `KNeighborsRegressor`, right?

Maybe I'm not seeing the full picture here and perhaps I'm missing the point of your comment.  If this package is completely agnostic to the type of `sklearn` or `sklearn`-derived estimator it supports, are you thinking that the spatial mode would support any "prediction-type" method that the estimator itself supports?  So `KNeighborsRegressor` implements both `predict` and `kneighbors`, but something like `LinearRegression` only supports `predict`.  So wouldn't you need a `kneighbors` methods in `sknnr-spatial` that is a spatial wrapper to `KNeighborsRegressor.kneighbors`, but it wouldn't be applicable to `LinearRegression`.  I don't think I'm explaining myself here well.

> If you're okay punting on this for now, I'd say we keep it in mind and revisit when we're a little further along and have a better sense of how compatible things are.

Yes, of course, definitely outside the scope of this PR.

> I think that'll probably be the case for `xarray`, but `rioxarray` and `rasterio` would probably only be needed if we add functionality for reading/writing rasters. Not sure if that's something you had in mind as part of `sknnr-spatial`, or if that's something we leave up to users? In any case, not something we need to make a decision on now!

You'll see this is a recurrent theme for me :wink:!  I always want to build more functionality into the package itself, whereas you wisely leave those types of decisions to the user.  Obviously, we can provide examples that incorporate `rioxarray`/`rasterio` (as you've done with your gist), but I agree that less might be more!

Reply:
>  If this package is completely agnostic to the type of sklearn or sklearn-derived estimator it supports, are you thinking that the spatial mode would support any "prediction-type" method that the estimator itself supports?

I haven't fully thought through how this would be implemented, but with an "estimator wrapper" approach, I think we could check which of our supported methods (i.e. `predict`, `kneighbors`, maybe `predict_proba`) exist on the estimator and wrap those with the necessary logic that handles flattening raster data from pixel-space to sample-space, filling NaNs, and chunking with Dask. That should give us a solution that works with `sklearn` or any compatible estimator package like `sknnr` via duck typing. This isn't a fully formed plan, so hopefully that makes sense and answers your question in broad strokes at least!

> I always want to build more functionality into the package itself, whereas you wisely leave those types of decisions to the user. Obviously, we can provide examples that incorporate rioxarray/rasterio (as you've done with your gist), but I agree that less might be more!

I'd probably call it strategic laziness rather than wisdom, but I'll take it! With that said, I'm not opposed to adding convenience features that would be helpful for users. The optimal balance might be some combo of that and examples showing integration with existing packages.